ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ResponseOutputStream.java

```diff
@@ -55,13 +59,29 @@ public void write(final @NotNull byte[] bytes, final int offset, final int lengt
     final byte[] bytesToWrite = new byte[length];
     System.arraycopy(bytes, offset, bytesToWrite, 0, length);
     final Buffer buffer = Buffer.buffer(bytesToWrite);
+    blockIfWriteQueueFull();
     response.write(buffer);
   }
 
   @Override
   public void close() {
     response.end();
   }
+
+  private void blockIfWriteQueueFull() {
+    VertxUtils.checkIsWorker();
+    if (response.writeQueueFull()) {
+      final CompletableFuture<Void> cf = new CompletableFuture<>();
+      response.drainHandler(v -> cf.complete(null));
+      try {
+        cf.get(60, TimeUnit.SECONDS);
+      } catch (Exception e) {
+        // Very slow consumers will result in a timeout, this will cause the push query to be closed
+        throw new KsqlException(e);
```

Comment:
Should we add a more informative error message?

Reply:
The wrapped exception should contain the important error messages. I don't usually like adding new messages on rethrows unless you're doing something extra.